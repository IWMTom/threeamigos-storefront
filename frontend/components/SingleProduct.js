import React, { Component } from "react";
import gql from "graphql-tag";
import { Query, Mutation } from "react-apollo";
import Head from "next/head";
import formatMoney from "../lib/formatMoney";
import { CURRENT_USER_QUERY } from "./User";
import { TOGGLE_CART_MUTATION } from "./Cart";

const SINGLE_PRODUCT_QUERY = gql`
	query SINGLE_PRODUCT_QUERY($slug: String, $first: Int = 1) {
		products(first: $first, where: { slug: $slug }) {
			id
			name
			image_url
			description
			price
			sale
			stock
			category {
				name
				icon
			}
			brand {
				name
			}
		}
	}
`;

const ADD_TO_CART_MUTATION = gql`
	mutation addToCart($id: ID!) {
		addToCart(id: $id) {
			id
			quantity
		}
	}
`;

class SingleProduct extends Component {
	render() {
		return (
			<Query
				query={SINGLE_PRODUCT_QUERY}
				variables={{
					slug: this.props.slug
				}}
			>
				{({ data, error, loading }) => {
					if (loading) return <p>Loading...</p>;
					if (error)
						return (
							<p data-test="graphql-error">
								Error: {error.message}
							</p>
						);
					const product = data.products[0];
					return (
						<React.Fragment>
							<Head>
								<title>{product.name} - Three Amigos</title>
							</Head>
							<div
								id="product-card"
								className={product.sale ? "card sale" : "card"}
							>
								{product.sale && (
									<span className="product-sale">
										Sale 50%
									</span>
								)}
								<div className="pure-g">
									<div className="pure-u-1-2">
										<img
											src={product.image_url}
											alt={product.name}
										/>
									</div>
									<div className="pure-u-1-2">
										<h1>{product.name}</h1>
										<div>
											<span id="category">
												<i
													className={
														product.category.icon
													}
												/>
												{product.category.name}
											</span>
											<span id="brand">
												<i className="ic_bookmark" />
												{product.brand.name}
											</span>
											<div className="clearfix" />
											<aside>
												<span>
													{product.sale && (
														<del>
															{formatMoney(
																product.price *
																	2
															)}
														</del>
													)}
													{formatMoney(product.price)}
												</span>
											</aside>
										</div>
										<p>{product.description}</p>
										<div>
											{product.stock > 0 && (
												<React.Fragment>
													<span className="stock-level">
														<strong>
															{product.stock}
														</strong>{" "}
														left in stock
													</span>

													<Mutation
														mutation={
															TOGGLE_CART_MUTATION
														}
													>
														{toggleCart => (
															<Mutation
																mutation={
																	ADD_TO_CART_MUTATION
																}
																variables={{
																	id:
																		product.id
																}}
																refetchQueries={[
																	{
																		query: CURRENT_USER_QUERY
																	}
																]}
															>
																{(
																	addToCart,
																	{ loading }
																) => (
																	<button
																		disabled={
																			loading
																		}
																		onClick={e => {
																			addToCart();
																			toggleCart();
																		}}
																		className="btn btn-pink"
																	>
																		Add
																		{loading &&
																			"ing"}{" "}
																		to Cart
																	</button>
																)}
															</Mutation>
														)}
													</Mutation>
												</React.Fragment>
											)}

											{product.stock == 0 && (
												<React.Fragment>
													<span className="stock-level no-stock">
														Out of stock
													</span>
													<button
														disabled="true"
														className="btn btn-pink"
													>
														Add to Cart
													</button>
												</React.Fragment>
											)}
										</div>
									</div>
								</div>
							</div>
						</React.Fragment>
					);
				}}
			</Query>
		);
	}
}

export default SingleProduct;
export { SINGLE_PRODUCT_QUERY };
