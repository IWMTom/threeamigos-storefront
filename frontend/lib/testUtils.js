import casual from "casual";

casual.seed(666);

const mockedProduct = () => ({
  __typename: "Product",
  id: "abcdefghijklmnopqrst",
  name: "Abc Def",
  image_url: "http://placehold.it/100x100",
  description: "dfghsdfidsfiosd",
  price: "199",
  sale: true,
  stock: 10,
  category: {
    name: "cat1",
    icon: "ic_bookmark"
  },
  brand: {
    name: "brand1"
  }
});

const mockedUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  photo_url: `${casual.word}.jpg`,
  cart: []
});

const mockedOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  total: 40000,
  products: [mockedOrderProduct(), mockedOrderProduct()]
});

const mockedOrderProduct = () => ({
  __typename: "OrderProduct",
  id: "abcdefghijklmnopqrst",
  slug: "abc-def",
  sale: true,
  name: "Abc Def",
  image_url: "http://placehold.it/100x100",
  price: "199",
  quantity: 1
});

const mockedCartProduct = overrides => ({
  __typename: "CartProduct",
  id: casual.uuid,
  quantity: 3,
  product: mockedProduct(),
  user: mockedUser(),
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  mockedProduct,
  mockedUser,
  mockedCartProduct,
  mockedOrder,
  mockedOrderProduct
};
