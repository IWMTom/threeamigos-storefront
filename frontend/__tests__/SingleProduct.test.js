import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import SingleProduct, {
	SINGLE_PRODUCT_QUERY
} from "../components/SingleProduct";
import { MockedProvider } from "react-apollo/test-utils";
import { mockedProduct } from "../lib/testUtils";

describe("<SingleProduct />", () => {
	it("renders with correct data", async () => {
		const mocks = [
			{
				request: {
					query: SINGLE_PRODUCT_QUERY,
					variables: { slug: "abc", first: 1 }
				},
				result: {
					data: {
						products: [mockedProduct()]
					}
				}
			}
		];
		const wrapper = mount(
			<MockedProvider mocks={mocks} addTypename={false}>
				<SingleProduct slug="abc" />
			</MockedProvider>
		);

		expect(wrapper.text()).toContain("Loading...");
		await wait();
		wrapper.update();
		expect(toJSON(wrapper.find("div#product-card"))).toMatchSnapshot();
	});

	it("errors when product isn't found", async () => {
		const mocks = [
			{
				request: {
					query: SINGLE_PRODUCT_QUERY,
					variables: { slug: "abc", first: 1 }
				},
				result: {
					errors: [{ message: "Product Not Found!" }]
				}
			}
		];

		const wrapper = mount(
			<MockedProvider mocks={mocks} addTypename={false}>
				<SingleProduct slug="abc" />
			</MockedProvider>
		);

		await wait();
		wrapper.update();

		const error = wrapper.find('[data-test="graphql-error"]');
		expect(toJSON(error)).toMatchSnapshot();
	});
});
