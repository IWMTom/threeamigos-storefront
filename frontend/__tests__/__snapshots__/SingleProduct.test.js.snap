// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> errors when product isn't found 1`] = `
<p
  data-test="graphql-error"
>
  Error: 
  GraphQL error: Product Not Found!
</p>
`;

exports[`<SingleProduct /> renders with correct data 1`] = `
<div
  className="card sale"
  id="product-card"
>
  <span
    className="product-sale"
  >
    Sale 50%
  </span>
  <div
    className="pure-g"
  >
    <div
      className="pure-u-1-2"
    >
      <img
        alt="Abc Def"
        src="http://placehold.it/100x100"
      />
    </div>
    <div
      className="pure-u-1-2"
    >
      <h1>
        Abc Def
      </h1>
      <div>
        <span
          id="category"
        >
          <i
            className="ic_bookmark"
          />
          cat1
        </span>
        <span
          id="brand"
        >
          <i
            className="ic_bookmark"
          />
          brand1
        </span>
        <div
          className="clearfix"
        />
        <aside>
          <span>
            <del>
              £3.98
            </del>
            £1.99
          </span>
        </aside>
      </div>
      <p>
        dfghsdfidsfiosd
      </p>
      <div>
        <span
          className="stock-level"
        >
          <strong>
            10
          </strong>
           
          left in stock
        </span>
        <Mutation
          mutation={
            Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": undefined,
                  "operation": "mutation",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [
                          Object {
                            "arguments": Array [],
                            "kind": "Directive",
                            "name": Object {
                              "kind": "Name",
                              "value": "client",
                            },
                          },
                        ],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "toggleCart",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 37,
                "start": 0,
              },
            }
          }
        >
          <Mutation
            mutation={
              Object {
                "definitions": Array [
                  Object {
                    "directives": Array [],
                    "kind": "OperationDefinition",
                    "name": Object {
                      "kind": "Name",
                      "value": "addToCart",
                    },
                    "operation": "mutation",
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [
                            Object {
                              "kind": "Argument",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "value": Object {
                                "kind": "Variable",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                              },
                            },
                          ],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "addToCart",
                          },
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "selections": Array [
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "quantity",
                                },
                                "selectionSet": undefined,
                              },
                            ],
                          },
                        },
                      ],
                    },
                    "variableDefinitions": Array [
                      Object {
                        "defaultValue": undefined,
                        "kind": "VariableDefinition",
                        "type": Object {
                          "kind": "NonNullType",
                          "type": Object {
                            "kind": "NamedType",
                            "name": Object {
                              "kind": "Name",
                              "value": "ID",
                            },
                          },
                        },
                        "variable": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                        },
                      },
                    ],
                  },
                ],
                "kind": "Document",
                "loc": Object {
                  "end": 81,
                  "start": 0,
                },
              }
            }
            refetchQueries={
              Array [
                Object {
                  "query": Object {
                    "definitions": Array [
                      Object {
                        "directives": Array [],
                        "kind": "OperationDefinition",
                        "name": undefined,
                        "operation": "query",
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "me",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "name",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "photo_url",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "cart",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "id",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "quantity",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "product",
                                          },
                                          "selectionSet": Object {
                                            "kind": "SelectionSet",
                                            "selections": Array [
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "id",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "price",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "image_url",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "name",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "slug",
                                                },
                                                "selectionSet": undefined,
                                              },
                                              Object {
                                                "alias": undefined,
                                                "arguments": Array [],
                                                "directives": Array [],
                                                "kind": "Field",
                                                "name": Object {
                                                  "kind": "Name",
                                                  "value": "stock",
                                                },
                                                "selectionSet": undefined,
                                              },
                                            ],
                                          },
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                        "variableDefinitions": Array [],
                      },
                    ],
                    "kind": "Document",
                    "loc": Object {
                      "end": 171,
                      "start": 0,
                    },
                  },
                },
              ]
            }
            variables={
              Object {
                "id": "abcdefghijklmnopqrst",
              }
            }
          >
            <button
              className="btn btn-pink"
              disabled={false}
              onClick={[Function]}
            >
              Add
               
              to Cart
            </button>
          </Mutation>
        </Mutation>
      </div>
    </div>
  </div>
</div>
`;
